File: php/src/Runtime.php
<?php

declare(strict_types=1);

namespace HyperUuid;

use FFI;

/**
 * FFI plumbing for the native libhyperuuid shared library — dlopen/dlsym plus a raw C-ABI
 * call, no runtime bridge (the same "no shim" positioning as the Go/Swift bindings' purego/
 * dlopen approach). PHP's built-in `ext-ffi` needs no Composer package for this — the same
 * "no extra dependency" stance as Go's purego-only/no-cgo approach.
 *
 * Performance shape, measured not assumed (the HyperCast lesson, applied here): PHP's raw
 * ext-ffi call floor is ~105 ns — already extension-class — so every avoidable nanosecond
 * in this file was wrapper, and the wrapper is written accordingly. Input byte strings are
 * declared `const char *` in the cdef so PHP strings cross zero-copy (no per-call CData
 * allocation, no memcpy in); the 16-byte out buffer is one static scratch allocated once
 * (PHP's request model makes static scratch safe — arrays auto-decay to pointers, so no
 * pre-taken address is needed); only the in-place byte-order rewrites still copy, because
 * the native call genuinely mutates the buffer.
 */
final class Runtime
{
    private static ?FFI $ffi = null;
    private static ?FFI\CData $out16 = null;

    public static function newV4(): string
    {
        $ffi = self::$ffi ?? self::load();
        $rc = $ffi->uuid_new_v4(self::$out16);
        if ($rc !== 0) {
            throw new RandomSourceException("uuid_new_v4 failed with code {$rc}");
        }
        return FFI::string(self::$out16, 16);
    }

    public static function newV5(string $namespaceBytes, string $nameBytes): string
    {
        $ffi = self::$ffi ?? self::load();
        $rc = $ffi->uuid_new_v5(
            $namespaceBytes,
            $nameBytes === '' ? null : $nameBytes,
            \strlen($nameBytes),
            self::$out16
        );
        if ($rc !== 0) {
            throw new RandomSourceException("uuid_new_v5 failed with code {$rc}");
        }
        return FFI::string(self::$out16, 16);
    }

    public static function newV6(int $unixMillis): string
    {
        $ffi = self::$ffi ?? self::load();
        $rc = $ffi->uuid_new_v6($unixMillis, self::$out16);
        if ($rc === 2) {
            throw new TimestampOutOfRangeException(
                'unix_millis does not fit the 60-bit v6 timestamp field'
            );
        }
        if ($rc !== 0) {
            throw new RandomSourceException("uuid_new_v6 failed with code {$rc}");
        }
        return FFI::string(self::$out16, 16);
    }

    public static function v6UnixMillis(string $bytes): int
    {
        $ffi = self::$ffi ?? self::load();
        return $ffi->uuid_v6_unix_millis($bytes);
    }

    public static function newV6Batch(int $count, int $unixMillis): string
    {
        if ($count === 0) {
            return '';
        }
        $ffi = self::$ffi ?? self::load();
        $out = $ffi->new('uint8_t[' . ($count * 16) . ']');
        $rc = $ffi->uuid_new_v6_batch($unixMillis, $count, $out);
        if ($rc === 2) {
            throw new TimestampOutOfRangeException(
                'unix_millis does not fit the 60-bit v6 timestamp field'
            );
        }
        if ($rc !== 0) {
            throw new RandomSourceException("uuid_new_v6_batch failed with code {$rc}");
        }
        return FFI::string($out, $count * 16);
    }

    public static function newV7(int $unixMillis): string
    {
        $ffi = self::$ffi ?? self::load();
        $rc = $ffi->uuid_new_v7($unixMillis, self::$out16);
        if ($rc === 2) {
            throw new TimestampOutOfRangeException(
                'unix_millis must fit within the RFC 9562 48-bit field'
            );
        }
        if ($rc !== 0) {
            throw new RandomSourceException("uuid_new_v7 failed with code {$rc}");
        }
        return FFI::string(self::$out16, 16);
    }

    public static function v7UnixMillis(string $bytes): int
    {
        $ffi = self::$ffi ?? self::load();
        return $ffi->uuid_v7_unix_millis($bytes);
    }

    public static function newV7Batch(int $count, int $unixMillis): string
    {
        if ($count === 0) {
            return '';
        }
        $ffi = self::$ffi ?? self::load();
        $out = $ffi->new('uint8_t[' . ($count * 16) . ']');
        $rc = $ffi->uuid_new_v7_batch($unixMillis, $count, $out);
        if ($rc === 2) {
            throw new TimestampOutOfRangeException(
                'unix_millis must fit within the RFC 9562 48-bit field'
            );
        }
        if ($rc !== 0) {
            throw new RandomSourceException("uuid_new_v7_batch failed with code {$rc}");
        }
        return FFI::string($out, $count * 16);
    }

    /**
     * Rewrites `$bytes` from RFC 9562 order to the byte order SQL Server's `uniqueidentifier`
     * needs on the wire to sort a version 7 UUID by creation order. Meaningful only for a
     * genuine version 7 UUID.
     */
    public static function v7ToSqlOrder(string $bytes): string
    {
        $ffi = self::$ffi ?? self::load();
        FFI::memcpy(self::$out16, $bytes, 16);
        $ffi->uuid_v7_to_sql_order(self::$out16);
        return FFI::string(self::$out16, 16);
    }

    /** Inverse of {@see v7ToSqlOrder} — rewrites `$bytes` from SQL Server order back to RFC 9562 order. */
    public static function v7ToRfcOrder(string $bytes): string
    {
        $ffi = self::$ffi ?? self::load();
        FFI::memcpy(self::$out16, $bytes, 16);
        $ffi->uuid_v7_to_rfc_order(self::$out16);
        return FFI::string(self::$out16, 16);
    }

    /**
     * Rewrites `$bytes` from RFC 9562 order to the byte order SQL Server's `uniqueidentifier`
     * needs on the wire to sort a version 6 UUID by creation order. Meaningful only for a
     * genuine version 6 UUID.
     */
    public static function v6ToSqlOrder(string $bytes): string
    {
        $ffi = self::$ffi ?? self::load();
        FFI::memcpy(self::$out16, $bytes, 16);
        $ffi->uuid_v6_to_sql_order(self::$out16);
        return FFI::string(self::$out16, 16);
    }

    /** Inverse of {@see v6ToSqlOrder} — rewrites `$bytes` from SQL Server order back to RFC 9562 order. */
    public static function v6ToRfcOrder(string $bytes): string
    {
        $ffi = self::$ffi ?? self::load();
        FFI::memcpy(self::$out16, $bytes, 16);
        $ffi->uuid_v6_to_rfc_order(self::$out16);
        return FFI::string(self::$out16, 16);
    }

    /**
     * Loaded lazily and exactly once, mirroring the Go binding's sync.Once / Swift's lazy
     * static let — the native library and its function pointers live for the process's
     * lifetime, same as every other binding (never unloaded).
     */
    private static function load(): FFI
    {
        [$rid, $libName] = NativePlatform::ridAndLibraryName();
        $path = __DIR__ . "/native/{$rid}/{$libName}";
        if (!is_file($path)) {
            throw new \RuntimeException(
                "hyperuuid: {$path} not found (unsupported platform, or this package was built "
                . 'without a native library for it)'
            );
        }

        self::$ffi = FFI::cdef(
            'int uuid_new_v4(void *out_ptr);'
            . 'int uuid_new_v5(const char *ns_ptr, const char *name_ptr, uint32_t name_len, void *out_ptr);'
            . 'int uuid_new_v6(uint64_t unix_millis, void *out_ptr);'
            . 'uint64_t uuid_v6_unix_millis(const char *uuid_ptr);'
            . 'int uuid_new_v6_batch(uint64_t unix_millis, uint32_t count, void *out_ptr);'
            . 'int uuid_new_v7(uint64_t unix_millis, void *out_ptr);'
            . 'uint64_t uuid_v7_unix_millis(const char *uuid_ptr);'
            . 'int uuid_new_v7_batch(uint64_t unix_millis, uint32_t count, void *out_ptr);'
            . 'void uuid_v7_to_sql_order(void *uuid_ptr);'
            . 'void uuid_v7_to_rfc_order(void *uuid_ptr);'
            . 'void uuid_v6_to_sql_order(void *uuid_ptr);'
            . 'void uuid_v6_to_rfc_order(void *uuid_ptr);',
            $path
        );
        self::$out16 = self::$ffi->new('uint8_t[16]');
        return self::$ffi;
    }
}

File: ruby/lib/hyperuuid/runtime.rb
require "fiddle"

module HyperUuid
  # Fiddle plumbing for the native libhyperuuid shared library — dlopen/dlsym plus a raw
  # C-ABI call, no runtime bridge (the same "no shim" positioning as the Go/Swift bindings'
  # purego/dlopen approach). Fiddle ships with every Ruby install; it's a plain gem
  # dependency here (see hyperuuid.gemspec) rather than a third-party one — mirroring Go's
  # "no cgo" and Python's zero-dependency PyO3 wheels.
  #
  # Unlike the Go/Swift bindings, which embed their native builds inside a single compiled
  # archive and must extract to a temp file before dlopen can see a real path, a Ruby gem's
  # files are already plain files on disk once installed — native/{rid}/{lib} can be
  # dlopen'd directly, no extraction step needed.
  module Runtime
    class RandomSourceError < StandardError; end
    class TimestampOutOfRangeError < StandardError; end

    NATIVE_DIR = File.join(__dir__, "native")

    @mutex = Mutex.new
    @functions = nil

    class << self
      def new_v4
        out = scratch
        rc = functions[:new_v4].call(out)
        raise RandomSourceError, "uuid_new_v4 failed with code #{rc}" unless rc.zero?
        out[0, 16]
      end

      def new_v5(namespace_bytes, name_bytes)
        out = scratch
        # Fiddle passes a String's bytes for void* directly (read-only) — no Pointer
        # wrapper, no copy — the same zero-copy crossing every other input here uses.
        name = name_bytes.empty? ? nil : name_bytes
        rc = functions[:new_v5].call(namespace_bytes, name, name_bytes.bytesize, out)
        raise RandomSourceError, "uuid_new_v5 failed with code #{rc}" unless rc.zero?
        out[0, 16]
      end

      def new_v6(unix_millis)
        out = scratch
        rc = functions[:new_v6].call(unix_millis, out)
        case rc
        when 0 then out[0, 16]
        when 2 then raise TimestampOutOfRangeError, "unix_millis does not fit the 60-bit v6 timestamp field"
        else raise RandomSourceError, "uuid_new_v6 failed with code #{rc}"
        end
      end

      def v6_unix_millis(bytes)
        functions[:v6_unix_millis].call(bytes)
      end

      def new_v6_batch(count, unix_millis)
        return "" if count.zero?
        out = Fiddle::Pointer.malloc(count * 16, Fiddle::RUBY_FREE)
        rc = functions[:new_v6_batch].call(unix_millis, count, out)
        case rc
        when 0 then out[0, count * 16]
        when 2 then raise TimestampOutOfRangeError, "unix_millis does not fit the 60-bit v6 timestamp field"
        else raise RandomSourceError, "uuid_new_v6_batch failed with code #{rc}"
        end
      end

      def new_v7(unix_millis)
        out = scratch
        rc = functions[:new_v7].call(unix_millis, out)
        case rc
        when 0 then out[0, 16]
        when 2 then raise TimestampOutOfRangeError, "unix_millis must fit within the RFC 9562 48-bit field"
        else raise RandomSourceError, "uuid_new_v7 failed with code #{rc}"
        end
      end

      def v7_unix_millis(bytes)
        functions[:v7_unix_millis].call(bytes)
      end

      def new_v7_batch(count, unix_millis)
        return "" if count.zero?
        out = Fiddle::Pointer.malloc(count * 16, Fiddle::RUBY_FREE)
        rc = functions[:new_v7_batch].call(unix_millis, count, out)
        case rc
        when 0 then out[0, count * 16]
        when 2 then raise TimestampOutOfRangeError, "unix_millis must fit within the RFC 9562 48-bit field"
        else raise RandomSourceError, "uuid_new_v7_batch failed with code #{rc}"
        end
      end

      def v7_to_sql_order(bytes)
        rewrite(:v7_to_sql_order, bytes)
      end

      def v7_to_rfc_order(bytes)
        rewrite(:v7_to_rfc_order, bytes)
      end

      def v6_to_sql_order(bytes)
        rewrite(:v6_to_sql_order, bytes)
      end

      def v6_to_rfc_order(bytes)
        rewrite(:v6_to_rfc_order, bytes)
      end

      # Whether this platform has a shared library for Fiddle to dlopen at all: a known RID
      # and the file actually present — in this install, or in the in-repo cargo build the
      # dev loop falls back to. Backend selection (hyperuuid.rb) asks this before falling
      # back to the WebAssembly backend, which needs neither.
      def fiddle_library_available?
        !library_path.nil?
      rescue NativePlatform::UnsupportedPlatformError
        false
      end

      private

      # The shared library to dlopen: this install's native/{rid}/{lib}, or — the
      # development loop — the in-repo cargo build, exactly what the other bindings' local
      # staging does (HyperCast's runtime has had this since its first release). Nil when
      # neither exists.
      def library_path
        rid, lib_name = NativePlatform.rid_and_library_name
        path = File.join(NATIVE_DIR, rid, lib_name)
        return path if File.exist?(path)

        repo_build = File.expand_path(File.join(__dir__, "../../../rust/target/release", lib_name))
        File.exist?(repo_build) ? repo_build : nil
      end

      # One 16-byte scratch allocation per thread, reused by every single-item call —
      # Fiddle::Pointer.malloc(..., RUBY_FREE) registers a GC finalizer per call, measured
      # (in HyperCast, same mechanism) as the dominant per-call cost by an order of
      # magnitude. Batches keep a per-call buffer: one malloc amortized over `count` IDs.
      def scratch
        Thread.current[:hyperuuid_scratch] ||= Fiddle::Pointer.malloc(16, Fiddle::RUBY_FREE)
      end

      # The in-place byte-order rewrites are the one shape that must copy in: the native
      # call genuinely mutates the buffer, and the input String is frozen.
      def rewrite(symbol, bytes)
        buf = scratch
        buf[0, 16] = bytes
        functions[symbol].call(buf)
        buf[0, 16]
      end

      # Loaded lazily and exactly once, mirroring the Go binding's sync.Once / Swift's lazy
      # static let — the native library and its function pointers live for the process's
      # lifetime, same as every other binding (never dlclose'd). The unsynchronized read is
      # the hot path; the mutex only guards the one-time load (a benign race — idempotent).
      def functions
        @functions || @mutex.synchronize { @functions ||= load_functions }
      end

      def load_functions
        path = library_path
        if path.nil?
          rid, lib_name = NativePlatform.rid_and_library_name
          raise LoadError,
                "hyperuuid: #{File.join(NATIVE_DIR, rid, lib_name)} not found (unsupported " \
                "platform, or this gem was built without a native library for it)"
        end

        handle = Fiddle.dlopen(path)
        {
          new_v4: Fiddle::Function.new(handle["uuid_new_v4"], [Fiddle::TYPE_VOIDP], Fiddle::TYPE_INT),
          new_v5: Fiddle::Function.new(
            handle["uuid_new_v5"],
            [Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP, Fiddle::TYPE_UINT32_T, Fiddle::TYPE_VOIDP],
            Fiddle::TYPE_INT
          ),
          new_v6: Fiddle::Function.new(
            handle["uuid_new_v6"],
            [Fiddle::TYPE_UINT64_T, Fiddle::TYPE_VOIDP],
            Fiddle::TYPE_INT
          ),
          v6_unix_millis: Fiddle::Function.new(
            handle["uuid_v6_unix_millis"],
            [Fiddle::TYPE_VOIDP],
            Fiddle::TYPE_UINT64_T
          ),
          new_v6_batch: Fiddle::Function.new(
            handle["uuid_new_v6_batch"],
            [Fiddle::TYPE_UINT64_T, Fiddle::TYPE_UINT32_T, Fiddle::TYPE_VOIDP],
            Fiddle::TYPE_INT
          ),
          new_v7: Fiddle::Function.new(
            handle["uuid_new_v7"],
            [Fiddle::TYPE_UINT64_T, Fiddle::TYPE_VOIDP],
            Fiddle::TYPE_INT
          ),
          v7_unix_millis: Fiddle::Function.new(
            handle["uuid_v7_unix_millis"],
            [Fiddle::TYPE_VOIDP],
            Fiddle::TYPE_UINT64_T
          ),
          new_v7_batch: Fiddle::Function.new(
            handle["uuid_new_v7_batch"],
            [Fiddle::TYPE_UINT64_T, Fiddle::TYPE_UINT32_T, Fiddle::TYPE_VOIDP],
            Fiddle::TYPE_INT
          ),
          v7_to_sql_order: Fiddle::Function.new(
            handle["uuid_v7_to_sql_order"],
            [Fiddle::TYPE_VOIDP],
            Fiddle::TYPE_VOID
          ),
          v7_to_rfc_order: Fiddle::Function.new(
            handle["uuid_v7_to_rfc_order"],
            [Fiddle::TYPE_VOIDP],
            Fiddle::TYPE_VOID
          ),
          v6_to_sql_order: Fiddle::Function.new(
            handle["uuid_v6_to_sql_order"],
            [Fiddle::TYPE_VOIDP],
            Fiddle::TYPE_VOID
          ),
          v6_to_rfc_order: Fiddle::Function.new(
            handle["uuid_v6_to_rfc_order"],
            [Fiddle::TYPE_VOIDP],
            Fiddle::TYPE_VOID
          ),
        }
      end
    end
  end
end
